clients/pkg/promtail/targets/kafka/consumer.go

```diff
@@ -0,0 +1,149 @@
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+)
+
+var defaultBackOff = backoff.Config{
+	MinBackoff: 1 * time.Second,
+	MaxBackoff: 60 * time.Second,
+	MaxRetries: 20,
+}
+
+type RunnableTarget interface {
+	target.Target
+	run()
+}
+
+type TargetDiscoverer interface {
+	NewTarget(sarama.ConsumerGroupSession, sarama.ConsumerGroupClaim) (RunnableTarget, error)
+}
+
+// consumer handle a group consumer instance.
+// It will create a new target for every consumer claim using the `TargetDiscoverer`.
+type consumer struct {
+	sarama.ConsumerGroup
+	discoverer TargetDiscoverer
+	logger     log.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	mutex          sync.Mutex // used during rebalancing setup and tear down
+	activeTargets  []target.Target
+	droppedTargets []target.Target
+}
+
+// start starts the consumer for a given list of topics.
+func (c *consumer) start(ctx context.Context, topics []string) {
+	c.wg.Wait()
+	c.wg.Add(1)
+
+	c.ctx, c.cancel = context.WithCancel(ctx)
+	level.Info(c.logger).Log("msg", "starting consumer", "topics", fmt.Sprintf("%+v", topics))
+
+	go func() {
+		defer c.wg.Done()
+		backoff := backoff.New(c.ctx, defaultBackOff)
+		for {
```

Comment:
should this be `backoff.ongoing`? It looks like it'll run as fast as possible as long as there's no error.

Reply:
The consume is blocking and no error when rebalance happens in which case we just want to re consume again